keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/nuts/Tree.java

```diff
@@ -0,0 +1,359 @@
+        this.acceptedLeapfrogCount = acceptedLeapfrogCount;
+        this.shouldContinueFlag = shouldContinueFlag;
+        this.deltaLikelihoodOfLeapfrog = deltaLikelihoodOfLeapfrog;
+        this.treeSize = treeSize;
+    }
+
+    /**
+     * A helper method to create a basic tree
+     *
+     * @param position                 the starting position
+     * @param momentum                 the starting momentum
+     * @param gradient                 the starting gradient
+     * @param initialLogOfMasterP      the initial log prob
+     * @param sampleAtAcceptedPosition the initial sample
+     * @return a basic tree
+     */
+    public static Tree createBasicTree(Map<VertexId, DoubleTensor> position,
+                                       Map<VertexId, DoubleTensor> momentum,
+                                       Map<VertexId, DoubleTensor> gradient,
+                                       double initialLogOfMasterP,
+                                       Map<VertexId, ?> sampleAtAcceptedPosition) {
+
+        return new Tree(
+            new Leapfrog(position, momentum, gradient),
+            initialLogOfMasterP,
+            sampleAtAcceptedPosition,
+            STARTING_ACCEPTED_LEAPFROG,
+            STARTING_SHOULD_CONTINUE,
+            STARTING_DELTA_LIKELIHOOD,
+            STARTING_TREE_SIZE
+        );
+    }
+
+    public static Tree buildOtherHalfOfTree(Tree currentTree,
+                                            List<Vertex<DoubleTensor>> latentVertices,
+                                            List<Vertex> probabilisticVertices,
+                                            LogProbGradientCalculator logProbGradientCalculator,
+                                            final List<? extends Vertex> sampleFromVertices,
+                                            double logU,
+                                            int buildDirection,
+                                            int treeHeight,
+                                            double epsilon,
+                                            double logOfMasterPMinusMomentumBeforeLeapfrog,
+                                            KeanuRandom random) {
+
+        Tree otherHalfTree = buildTree(
+            latentVertices,
+            probabilisticVertices,
+            logProbGradientCalculator,
+            sampleFromVertices,
+            buildDirection == -1 ? currentTree.leapfrogBackward : currentTree.leapfrogForward,
+            logU,
+            buildDirection,
+            treeHeight,
+            epsilon,
+            logOfMasterPMinusMomentumBeforeLeapfrog,
+            random
+        );
+
+        if (buildDirection == -1) {
+            currentTree.leapfrogBackward = otherHalfTree.leapfrogBackward;
+        } else {
+            currentTree.leapfrogForward = otherHalfTree.leapfrogForward;
+        }
+
+        return otherHalfTree;
+    }
+
+    private static Tree buildTree(List<Vertex<DoubleTensor>> latentVertices,
```

Comment:
Earlier I suggested pulling the builder out into a separate class, and you can put most of these arguments into its constructor to make it clearer what's invariant and what's not.